### ISSUE TITLE
AH finder diverges sometimes

### ISSUE THREAD

**Issue by nilsvu** [sxs-collaboration/spectre#3899]
# Bug reports:

I created a directory on Wheeler in `/home/nfischer/2022-debug-ah-finder/` that reproduces a bug with the AH finder.

The directory contains volume data on a single time slice (the result of an initial data solve) in the `BbhVolume*.h5` files. I find horizons in the data using the standalone `FindHorizons3D` executable on the branch:

- https://github.com/nilsvu/spectre/tree/find_binary_horizons

Running the horizon finder on varying number of cores and nodes produces very different results, and sometimes diverges. See the `run*/` subdirectories for example runs on Minerva.

To reproduce:

1. Compile `FindHorizons3D` on the branch linked above.
2. Point the `Importers.VolumeData.FileGlob` in the `FindHorizons.yaml` input file to the `BbhVolume*.h5` files on your system.
3. Run `FindHorizons3D` with the `FindHorizons.yaml` input file on different numbers of cores and nodes. On Minerva, the AH finder succeeded on 120 cores (8 nodes) but diverged on 240 cores (16 nodes).

**Comment by geoffrey4444** [comment]
`ocean:/home/geoffrey/aps2022/ahdebug/nilsvu/2022-debug-ah-finder/12x20/lotsoruns` has 100 identical runs (same submission script, same yaml file) using the steps above, except using @nilsvu `xcts_spins_test` branch because `find_binary_horizons` gives an option parsing error.

So far, of 29 complete runs, 17 failed with diverging apparent-horizon finds, while 12 succeeded.

For those runs that succeed, the horizon quantities (irreducible mass, min/max Ricci, etc.) are consistent to several digits):

```
AhA results:
(0,0): 0, 24.4586, 0.697559, 2.14708, -0.34025, 0.797197, 0.847181
   (0,0): 0, 24.459, 0.697564, 2.14701, -0.340198, 0.7972, 0.847176
   (0,0): 0, 24.4588, 0.697562, 2.14698, -0.340043, 0.7972, 0.847181
   (0,0): 0, 24.4588, 0.697562, 2.14698, -0.339996, 0.7972, 0.847181
   (0,0): 0, 24.4586, 0.697559, 2.14708, -0.340298, 0.797196, 0.847181
   (0,0): 0, 24.4589, 0.697562, 2.14698, -0.339993, 0.7972, 0.847181
   (0,0): 0, 24.4586, 0.697559, 2.14708, -0.3403, 0.797197, 0.847181
   (0,0): 0, 24.4586, 0.697559, 2.14708, -0.3403, 0.797197, 0.847181
   (0,0): 0, 24.4588, 0.697562, 2.14698, -0.339997, 0.7972, 0.847182
   (0,0): 0, 24.4588, 0.697561, 2.14711, -0.340454, 0.797197, 0.847177
   (0,0): 0, 24.4589, 0.697563, 2.14698, -0.339992, 0.7972, 0.847181
```

```
AhB results:
(0,0): 0, 5.11244, 0.318918, 4.94364, 4.89742, 0.318925, 0.0126695
   (0,0): 0, 5.11244, 0.318918, 4.94364, 4.89735, 0.318925, 0.0126686
   (0,0): 0, 5.11244, 0.318918, 4.94364, 4.89738, 0.318925, 0.0126688
   (0,0): 0, 5.11244, 0.318918, 4.9436, 4.89736, 0.318925, 0.0126691
   (0,0): 0, 5.11244, 0.318918, 4.94359, 4.8974, 0.318925, 0.0126654
   (0,0): 0, 5.11244, 0.318918, 4.94359, 4.89741, 0.318925, 0.012668
   (0,0): 0, 5.11244, 0.318918, 4.9436, 4.89738, 0.318925, 0.0126679
   (0,0): 0, 5.11244, 0.318918, 4.94355, 4.89741, 0.318925, 0.0126688
   (0,0): 0, 5.11244, 0.318918, 4.94359, 4.89735, 0.318925, 0.0126682
   (0,0): 0, 5.11244, 0.318918, 4.94364, 4.89738, 0.318925, 0.0126688
   (0,0): 0, 5.11244, 0.318918, 4.94359, 4.89745, 0.318925, 0.0126664
   (0,0): 0, 5.11244, 0.318918, 4.94359, 4.89734, 0.318925, 0.0126697
   (0,0): 0, 5.11244, 0.318918, 4.94355, 4.89745, 0.318925, 0.0126675
   (0,0): 0, 5.11244, 0.318918, 4.9436, 4.89741, 0.318925, 0.0126675
   (0,0): 0, 5.11244, 0.318918, 4.9436, 4.89736, 0.318925, 0.0126691
   (0,0): 0, 5.11244, 0.318918, 4.9436, 4.89738, 0.318925, 0.0126679
   (0,0): 0, 5.11244, 0.318918, 4.94354, 4.89742, 0.318925, 0.0126685
   (0,0): 0, 5.11244, 0.318918, 4.94364, 4.89743, 0.318925, 0.0126681
   (0,0): 0, 5.11244, 0.318918, 4.9436, 4.89738, 0.318925, 0.0126679
   (0,0): 0, 5.11244, 0.318918, 4.94355, 4.89742, 0.318925, 0.0126696
   (0,0): 0, 5.11244, 0.318918, 4.94359, 4.89742, 0.318925, 0.0126665
   (0,0): 0, 5.11244, 0.318918, 4.94359, 4.89744, 0.318925, 0.0126664
```

**Comment by markscheel** [comment]
Does the problem still occur if there is only one horizon and not two?

**Comment by markscheel** [comment]
What is missing from `find_binary_horizons`?  Is it easy to add that commit?

**Comment by markscheel** [comment]
Ok, I can reproduce this on wheeler now, with `xcts_spins_test` and cherry-picking Nils D.'s recent changes to wheeler modules.   No failures for 4 nodes or 8 nodes so far.  Sometimes failures with 12 nodes.

**Comment by geoffrey4444** [comment]
`ocean:/home/geoffrey/aps2022/ahdebug/nilsvu/2022-debug-ah-finder/12x20/lotsoruns1AH` reproduces after by hand removing AhB from FindHorizons.hpp: 17 diverge on 12 20-core nodes, 13 succeed, out of 30 attempts

**Comment by nilsvu** [comment]
I pushed an update to the `find_binary_horizons` branch with only the essential commits, including the Wheeler env update.

**Comment by markscheel** [comment]
I can reproduce the problem with one horizon.  Also, I noticed that the divergence is mild, and looks more like there is a source of noise rather than a horrible bug.  Even in the cases where the AH finder converges, the residual sometimes jumps around and sometimes increases temporarily, so it takes more iterations than it 'should'.

So I then tried offsetting the expansion center of the AH finder slightly (I used `Center: [-3.87003, 0.00002, 0.00001]`). And so far it converges every time, even if I reduce the tolerance to `TruncationTol: 0.0001`.

Offsetting the expansion center eliminates Strahlkorper points that happen to lie exactly on the boundaries between multiple Elements.  (With the original expansion center, there are many such points because of symmetry; for example every point on the equator of the Strahlkorper is such a point, as is every point at angle phi=0 on the Strahlkorper).

So it appears that the problem is noise from these boundary points.  For boundary points, the first Element that interpolates to that point wins, i.e. the InterpolationTarget fills that point and is done with it, and it subsequently ignores other Elements that try to fill that same point.  But which element is "first" depends on the random order of messages, and hence there is noise.

I predict that this problem should get worse as the data becomes less smooth (because two neighboring Elements will disagree more at boundary points), and likewise it should get worse as the AH finder tolerance decreases.  And the problem gets worse with more cores because there are more messages that arrive in a different order.

It should be possible to redo InterpolationTargets so that they store even more data (i.e. they would store duplicate data sets for the same interpolation point) and then average over duplicate data after all messages have arrived.  This will require that InterpolationTargets wait for all communications before proceeding (that doesn't happen now; if an InterpolationTarget has all its points filled it just finishes and calls the post_interpolation_callback, and extra communications that it receives later that are tagged for the now-completed interpolation are ignored).

**Comment by geoffrey4444** [comment]
Interesting, @markscheel ! How involved do you think it would be to make such a change?

**Comment by markscheel** [comment]
I am working on such a change now.  It's not a trivial thing like making `post_horizon_find_callback` a list, because it changes how the logic works a bit, and it requires initializing/using/cleaning-up new things in the InterpolationTarget's DataBox, but hopefully it won't be too bad.

**Comment by markscheel** [comment]
Ugh. This change will result in many more messages.

Currently if an Element has zero Strahlkorper points (which is the case for most Elements), it doesn't send a message to the InterpolationTarget.  The InterpolationTarget receives messages until all its Strahlkorper points are filled, and then it finishes and ignores subsequent messages if any occur (i.e. points shared by more than one Element). 

But now every Element (more precisely, every Interpolator element not every DG element) needs to send a message to the InterpolationTarget, because otherwise the InterpolationTarget doesn't know when it can finish; it doesn't know if it has duplicate data until it has heard from everyone.

**Comment by geoffrey4444** [comment]
Thanks @markscheel !! I really appreciate your working on this.  Meanwhile, I verified that offsetting the AH finders by O(1e-5) worked on ocean as well: the 1-AH FindHorizons3D test above not only succeeded every time, but the output in Horizons.h5 was identical each time.

**Comment by knelli2** [comment]
@markscheel I'm confused, but maybe I don't understand this that well. If most Elements don't have Strahlkorper points and thus don't send messages to the InterpolationTarget, is there some way to let the InterpolationTarget know that it's only going to be receiving messages from a certain subset of the Elements? (i.e. the ones *with* Strahlkorper points) And then the InterpolationTarget just was to wait to receive messages from all *those* elements, even if some of the points are duplicated?

**Comment by nilsvu** [comment]
@markscheel great that you were able to reproduce it and that you found the root cause! Here's an idea:

@knelli2 suggests a smarter registration, which can be nontrivial with a time-dependent domain and with AMR, so I think I understand why you did the "first received message wins" approach. However, we have one thing going for us here: the AH moves around and deforms along with the excision, so in practice the set of elements contributing to it will remain approximately the same. So in most timesteps we shouldn't have to wait for all other elements to confirm that they haven't changed. One way of handling this that I can think of is we only wait for registration messages from unregistered elements if one of the currently registered elements sent a "I'm not part of the interpolation surface anymore" message instead of sending data. Then, usually the AH finder only waits for the set of registered elements (all that intersect with the interpolation surface), and only when one of them stops intersecting does the AH finder wait for other elements to register and "fill the void".

**Comment by markscheel** [comment]
@nilsvu @knelli2 yes I think we eventually need smarter (re)registration.    The difficulty is that the AH surface *does* move around *from iteration to iteration*. So even from one iteration to the next (at the same time), Elements need to be turned on and off. So there's a lot of bookkeeping involved and perhaps even more messages.  And we also need to handle the case where an Element has multiple points and only one point moves to a new Element (or even worse, the point moves to the boundary of the same Element and is now shared); in this case a new Element needs to activate but the old one still remains active.

I agree that something like this is what we want in the end, but my plan is to not do it in the current fix, so that we have an improved horizon finder in a matter of days rather than weeks.

**Comment by markscheel** [comment]
By the way, SpEC does something similar. In SpEC,  all Strahlkorper-containing Elements *and their nearest neighbors* send messages.  This might not be enough for SpECTRE because SpECTRE elements are smaller than SpEC elements (so greater chance that the horizon moves by >1 element in an iteration).

**Comment by markscheel** [comment]
It was decided on a SpECTRE call March 31 to implement an alternate approach:  @kidder  will add to Elements the ability to know their edge and corner neighbors.  Then before interpolating to a point, each Element (or the the Interpolator element associated with that DG element) will check if that point is on a boundary, and if so, it will determine the `ElementId`s of all the neighbors that can interpolate that point.  The element with the smallest `ElementId` will do the interpolation on that point and send the message to the InterpolationTarget, and the other elements will ignore that point.

This has the advantage that (like the current setup) there is no need for InterpolationTargets to know ahead of time how many Elements or (Interpolator elements) will send it messages.

**Comment by knelli2** [comment]
I just tried adding @nilsvu's branch on top of the latest develop (58d7cf67f53631934b4ec7067762f189e53ebc17) which has #3913 merged. Here is the branch https://github.com/knelli2/spectre/tree/check_horizon_finder.

I tested this on wheeler at `wheeler:/home/knelli/ds09/test_ah_finder`. I did a total of 48 runs: 24 runs on 4 nodes (96 cores) and 24 runs on 6 nodes (144 cores). In this directory there are two directories `4node_runs` and `6node_runs` that hold all the runs (named `try*`). Each run directory has files like `horizon_A_out.txt` and `horizon_B_out.txt` where I separated the output from AhA and AhB to make it easier to compare. Also in the `test_ah_finder` directory, I have files `compare_A.txt` and `compare_B.txt` where I `diff` the `horizon_?_out.txt` file from one run with each of the other runs.

For the most part, the output of all runs was completely identical. Same number of its, and the residuals printed to stdout were identical (to whatever precision they were printed out to). However, a few runs failed, but I'm not sure it's related to this issue of the ordering of messages received by the interpolator:

- `4node_runs/try18` failed with
```
*** Error in `/home/knelli/ds09/test_ah_finder/4node_runs/try18/FindHorizons3D': malloc(): memory corruption: 0x00007f3668028e20 ***
```

- `6node_runs/try12` failed with
```
------------- Processor 92 Exiting: Caught Signal ------------
Reason: Segmentation fault
```

- `6node_runs/try21` failed with
```
*** Error in `/home/knelli/ds09/test_ah_finder/6node_runs/try21/FindHorizons3D': corrupted double-linked list: 0x00007f65a42c0a90 ***
```

This may be due to the false quiescence issue @geoffrey4444 and I have occasionally seen. In all three of these errors, either one or both of the horizon finders failed to start (or at least didn't finish a first iteration to produce any output). But this time, an error occured instead of just exiting.

So other than these 3 errors, the horizon finder seems to be consistently reproducible with @kidder's changes.
